i.MX RT networking fixes after SDK 2.5.0 update

agansari opened the pull request:
Work in progress, used for reference in pull #12457

aurel32 commented:
@agansari nocache memory regions are currently not supported with the NXP MPU, only the ARMv7 MPU is supported.

That said, I am not convinced you actually need to define the TX and RX buffers as nocache. I believe that the SDK 2.5.0 changed the Ethernet driver to use DMA. In that case, as long as a single cache line of the buffer can not simultaneously be read and written by both the CPU and the device, you should use cache operations instead:
- Before reading data from the RX buffer and after the DMA operation ended, you should _invalidate_ the memory range corresponding to the buffer with `SCB_InvalidateDCache_by_Addr`.
- After writing data to the TX buffer and before starting the DMA operation, you should _clean_ the memory range corresponding to the buffer with `SCB_CleanDCache_by_Addr`.

The buffers need to be aligned to at least the size of a cache line, which seems to be the case already.

aurel32 commented:
Hmm, I have just looked at `ext/hal/nxp/mcux/drivers/imx/fsl_enet.c`and it seems that enabling `FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL` already handle the cache clean/invalidate operations.

MaureenHelm commented:
> @agansari nocache memory regions are currently not supported with the NXP MPU, only the ARMv7 MPU is supported.

i.MX RT parts have the ARMv7 MPU, not the NXP MPU.

MaureenHelm reviewed (commented):
> Hmm, I have just looked at `ext/hal/nxp/mcux/drivers/imx/fsl_enet.c`and it seems that enabling `FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL` already handle the cache clean/invalidate operations.

I think @aurel32 	is right.

aurel32 commented:
> > @agansari nocache memory regions are currently not supported with the NXP MPU, only the ARMv7 MPU is supported.
> 
> i.MX RT parts have the ARMv7 MPU, not the NXP MPU.

Oh right. Given it's an NXP chip, I had in mind it was a NXP MPU. Then nocache should be usable, although that might not be the best solution.

aurel32 commented:
I have been able to get the ethernet working on my IMX.RT1020 board, based on the RT1050 code, some PLL changes, and the PHY changes from this PR. The PHY changes looks ok to me and seems to work fine (at least the RT1020 version). I guess it might be a good idea to move those changes into a separate commit.

I have tried to reproduce the cache coherency issues on this board. I have not been able to get SDRAM0 working (but I haven't tried hard), so i switched to OCRAM0 instead. It indeed broke the ethernet driver. So far I have tested that:
- Enabling `CONFIG_DEVICE_CONFIGURATION_DATA` and `FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL` with DTCM0: the ethernet driver stopped working
- Enabling `CONFIG_DEVICE_CONFIGURATION_DATA` and `FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL` with OCRAM0: the ethernet driver doesn't work
- Using nocache for the buffers: the ethernet driver doesn't work

It's not very successful, I'll try to look again at it tomorrow night. That said I think `CONFIG_DEVICE_CONFIGURATION_DATA` is not the good condition to enable the cache operations on the buffer, as OCRAM0 does not need it for example.

aurel32 commented:
> - Enabling CONFIG_DEVICE_CONFIGURATION_DATA and FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL with DTCM0: the ethernet driver stopped working

I probably did something wrong when testing, I am not able to reproduce that anymore. So just ignore.

aurel32 commented:
Ok, I have made some progress to the point the ethernet driver now works with OCRAM0, though it's still a bit hackish. I realized that there are actually buffers and buffer descriptors. The first ones are handled through cache operations (by enabling `FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL`), and the second ones by defining the region as `__nocache`. This is actually quite similar to what is done in PR #11888.

Now, we have decided to not define the nocache memory region as strongly ordered, so there is still need to add a __DMB() after writing to the nocache memory region. It looks like that those are missing in the MCUX driver. This can be workarounded by changing the nocache region to strong ordering:

```
--- a/arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h
+++ b/arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h
@@ -122,7 +122,7 @@ static inline void _get_mpu_ram_nocache_region_attr(arm_mpu_region_attr_t *p_att
         */
        (void) base;
 
-       p_attr->rasr = _get_region_attr(1, ap, 1, 0, 0, 1, 0, size);
+       p_attr->rasr = _get_region_attr(1, ap, 0, 0, 0, 1, 0, size);
 }
 #endif /* CONFIG_NO_CACHE_MEMORY */
```
 
That said it's probably better to add the missing memory barriers in the MCUX driver. Anyway bed time.

agansari commented:
@MaureenHelm will remove flash/sdram increase. Also FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL is enabled when using SDRAM with DCD, I'm not doing any other management outside fsl_enet for networking.
The problem I currently have is: when I flash the board with eth enabled then power off via unplug and power it up via USB, it gets stuck. Running non-eth samples board boots okay.

@aurel32 good to hear you enabled networking on 1020 👍 
If you run on DTCM/OCRAM you don't need _nocache or cache coherency functions enabled by FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL (but should not affect). The cache coherency is automatically managed inside the ARM core, before @MaureenHelm enabled SDRAM, i used to run on DTCM and didn't have coherency problems. Adding an external RAM that is outside auto-coherency domain will require manual cache management or we risk having incoherence between the core and different subsystems.

aurel32 commented:
> If you run on DTCM/OCRAM you don't need _nocache or cache coherency functions enabled by FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL (but should not affect). The cache coherency is automatically managed inside the ARM core, before @MaureenHelm enabled SDRAM, i used to run on DTCM and didn't have coherency problems.

I disagree with that, the ARM core doesn't manage coherency, it has to be done in software. What happens is that the DTCM memory is directly coupled to the ARM core and thus bypasses the cache. Accesses to the OCRAM is done through the cache, and thus both `FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL` and `__nocache` are needed. I have confirmed that on my board.

aurel32 commented:
> @MaureenHelm will remove flash/sdram increase. Also FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL is enabled when using SDRAM with DCD, I'm not doing any other management outside fsl_enet for networking.
The problem I currently have is: when I flash the board with eth enabled then power off via unplug and power it up via USB, it gets stuck. Running non-eth samples board boots okay.

Have you tried the patch changing `_get_mpu_ram_nocache_region_attr`? Without this patch, it sometimes work, sometimes not, it's a bit random and can match the behaviour you observed.

agansari commented:
> Have you tried the patch changing `_get_mpu_ram_nocache_region_attr`? Without this patch, it sometimes work, sometimes not, it's a bit random and can match the behaviour you observed.

What patch are you refering to: _get_mpu_ram_nocache_region_attr ?

aurel32 commented:
> What patch are you refering to: _get_mpu_ram_nocache_region_attr ?

The one below, changing the nocache region to strongly ordered. For me it's necessary to get ethernet working reliably. It's just a workaround, I think the correct way to fix that is to add the missing __DMB() in the MCUX driver:

```
--- a/arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h
+++ b/arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h
@@ -122,7 +122,7 @@ static inline void _get_mpu_ram_nocache_region_attr(arm_mpu_region_attr_t *p_att
         */
        (void) base;
 
-       p_attr->rasr = _get_region_attr(1, ap, 1, 0, 0, 1, 0, size);
+       p_attr->rasr = _get_region_attr(1, ap, 0, 0, 0, 1, 0, size);
 }
 #endif /* CONFIG_NO_CACHE_MEMORY */

```

aurel32 commented on boards/arm/mimxrt1050_evk/Kconfig.defconfig:
```diff
@@ -10,6 +10,15 @@ if BOARD_MIMXRT1050_EVK || BOARD_MIMXRT1050_EVK_QSPI
+config DEVICE_CONFIGURATION_DATA
+	default y
+
+config NOCACHE_MEMORY
```
`NOCACHE_MEMORY` is user configurable, so if a user disables it through menuconfig, the ethernet driver won't work. I think it should be better to change `drivers/ethernet/Kconfig.mcux` to:

```
menuconfig ETH_MCUX
        bool "MCUX Ethernet driver"
        depends on HAS_MCUX
        select NOCACHE_MEMORY if CPU_CORTEX_M7
        select ARM_MPU if CPU_CORTEX_M7
```

This way, it ensures the ethernet driver always work.

aurel32 commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -107,11 +107,19 @@ struct eth_context {
 
 static void eth_0_config_func(void);
 
+#ifndef CONFIG_NOCACHE_MEMORY
```
I think there is no need for #ifdef there. __nocache should always be used as it has no performance impact when DTCM is used.

aurel32 commented on ext/hal/nxp/mcux/drivers/imx/CMakeLists.txt:
```diff
@@ -10,8 +10,13 @@ zephyr_library_compile_definitions_ifdef(
 
+zephyr_library_compile_definitions_ifdef(
+  CONFIG_DEVICE_CONFIGURATION_DATA FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL
+)
```
I think this should be enabled unconditionaly, so that OCRAM is also supported. The impact with DTCM should be minimal.

aurel32 commented on ext/hal/nxp/mcux/drivers/imx/CMakeLists.txt:
```diff
@@ -10,8 +10,13 @@ zephyr_library_compile_definitions_ifdef(
 zephyr_sources_ifdef(CONFIG_UART_MCUX_LPUART  fsl_lpuart.c)
 zephyr_sources_ifdef(CONFIG_ETH_MCUX          fsl_enet.c)
 zephyr_sources_ifdef(CONFIG_ENTROPY_MCUX_TRNG fsl_trng.c)
+zephyr_sources_ifdef(CONFIG_DEVICE_CONFIGURATION_DATA fsl_cache.c)
```
same here.

aurel32 commented:
> Thanks for the new version. The two `__DMB()` improves the situation a bit, but I can get it working reliably. For example `samples/net/sockets/dumb_http_server` sometimes hang when downloading the data.

The `__DMB()` instruction has to be placed when the changes done to the nocache region before this instruction should be visible from the instructions after. I have checked the MCUX code, and it seems it mostly happens when calling `ENET_ActiveSend()`. `ENET_SetTxBufferDescriptors` and `ENET_SetRxBufferDescriptors` also write the nocache region, but given it is done way before the device is enabled, so that is not a problem. I therefore came with the following patch which make the ethernet device to work correctly on my imxrt1020 board:

```
--- a/ext/hal/nxp/mcux/drivers/imx/fsl_enet.c
+++ b/ext/hal/nxp/mcux/drivers/imx/fsl_enet.c
@@ -886,6 +888,8 @@ static void ENET_ActiveSend(ENET_Type *base, uint32_t ringId)
 {
     assert(ringId < FSL_FEATURE_ENET_QUEUE);
 
+    __DMB();
+
     switch (ringId)
     {
         case kENET_Ring0:
```

agansari commented:
@aurel32 I would not change fsl_enet.c for maintenance purposes. I would use the barrier right before ENET_SendFrame in eth_mcux.c, the problem is I see there is some memcpy inside the driver.

I'll make the changes you requested regarding NOCACHE_MEMORY configuration.

aurel32 commented:
> @aurel32 I would not change fsl_enet.c for maintenance purposes.

> I would use the barrier right before ENET_SendFrame in eth_mcux.c, the problem is I see there is some memcpy inside the driver.

Putting the barrier before `ENET_SendFrame` is useless as the write to the TX buffer descriptor actually happen in this function. The barrier has to be after the write and before activating the buffer descriptor.

agansari commented:
NXP skd engineers can make this change, but i'm not sure it completely fixes the issue.
I'm also testing dumb_http_server on 1050 and it's a lot slower than previously. I'm looking back at code changes and setup and checking performance using:

> ab -n1000 http://192.0.2.1:8080/

aurel32 commented:
> NXP skd engineers can make this change, but i'm not sure it completely fixes the issue.

For me it does, I haven't notice any reliability issues.

> I'm also testing dumb_http_server on 1050 and it's a lot slower than previously. I'm looking back at code changes and setup and checking performance using:
> 
> > ab -n1000 http://192.0.2.1:8080/

Indeed, I confirm that the network interface is quite slow, I only get 5.21 req/sec on my RT1020 board vs 110.61 req/sec on the SAM E70 board. That said I also tried with DTCM instead of OCRAM, and the performance is unchanged. I have tried to use the ITCM instead of the QSPI flash, but I haven't been manage to reduce my configuration enough for the code to fit.

agansari commented:
@aurel32 with the latest changes on 1050 i get ~160 req/s, cam you check on your side with 1020?
__DMB() was replaced by the lesser __DSB() (data only sync).

@MaureenHelm setting rx and tx buffers (not descriptors) to __nocache reproduce issue #3132 on rt1050.

aurel32 commented:
> NXP skd engineers can make this change, but i'm not sure it completely fixes the issue.

@ansgari You were correct. I have found another place with a missing barrier, here is the corresponding patch:

```
--- a/ext/hal/nxp/mcux/drivers/imx/fsl_enet.c
+++ b/ext/hal/nxp/mcux/drivers/imx/fsl_enet.c
@@ -1477,6 +1477,8 @@ static void ENET_UpdateReadBuffers(ENET_Type *base, enet_handle_t *handle, uint3
         handle->rxBdCurrent[ringId]++;
     }
 
+    __DSB();
+
     /* Actives the receive buffer descriptor. */
     switch (ringId)
     {
```

Basically the idea is to put a barrier before activating the buffer descriptors to make sure they are completely flushed out to the RAM. They are activated with the `RDAR` or `TDAR `registers, so the best is to look for them in the code. With the above patch things works much more reliably, although it doesn't change anything with regard to performance. I'll continue debugging that. Note that I have found that this barrier is actually needed for DTCM, OCRAM and SDRAM (yes I got this one working!).

I also looked more in details about the latest version of your PR. Thanks for the update it looks almost ready. I have just noticed that the driver also applies to other SoCs with Cortex-M4. I asked you to drop the version without __nocache, but given it is not needed on Cortex-M4, it might have been a bad idea, I am very sorry about that. Feel free to revert that part to the previous version and enable __nocache only on the cortex M7. I have also seen that some of the changes ended-up in the wrong commit, so the content of each patch doesn't fully match the description anymore.

aurel32 commented:
@agansari I found the reason of the TCP slowness, it seems the TRNG is limit to one number every 60ms, at least on my board. I don't know if it is normal or a driver issue, as I can't find any documentation about that. The TCP stack needs random numbers to generate the sequence number and the source port.

After setting `CONFIG_ENTROPY_DEVICE_RANDOM_GENERATOR=n` and `CONFIG_XOROSHIRO_RANDOM_GENERATOR=y`, I get ~115 req/s. Removing the printf from the code, I even get ~360 req/s.

aurel32 commented:
> @MaureenHelm setting rx and tx buffers (not descriptors) to __nocache reproduce issue #3132 on rt1050.

@jukkar confirmed on Slack that adding the two memory barriers to the kinetis enet driver fixes the issue.

agansari commented:
> @jukkar confirmed on Slack that adding the two memory barriers to the kinetis enet driver fixes the issue.

This does not work on my side. I can still reproduce 1000ms pings. Still looks like a driver/stack flooding/interaction issue. 
Hw RNG delays may also slow down the stack as you pointed out.

aurel32 commented on drivers/ethernet/Kconfig.mcux:
```diff
@@ -7,6 +7,8 @@
 menuconfig ETH_MCUX
 	bool "MCUX Ethernet driver"
 	depends on HAS_MCUX
+	select NOCACHE_MEMORY if CPU_CORTEX_M7
```
Minor nitpick, you might now want to use `HAS_MCUX_CACHE` instead of `CPU_CORTEX_M7`. Otherwise it looks all good.

jukkar commented:
> @jukkar confirmed on Slack that adding the two memory barriers to the kinetis enet driver fixes the issue.

Note that I did not do my testing with ping, but by flooding zephyr with UDP packets as described in my comment at https://github.com/zephyrproject-rtos/zephyr/issues/3132#issuecomment-446242662
After enabling the memory barrier in the kinetis driver, I did not see various crashes even with 2 RX buffers.

pfalcon commented:
Trying this on top of https://github.com/zephyrproject-rtos/zephyr/commit/26c9c7480ec0a379809e3ed7fdf368c6446edba7 , I 100% reproducibly get:

~~~
***** Booting Zephyr OS zephyr-v1.13.0-3393-gdd987f07de *****


[00:00:00.011,625] <inf> eth_mcux: Enabled 100M full-duplex mode.
[00:00:00.058,796] <inf> net_config: Initializing network
[00:00:00.058,820] <inf> net_config: IPv4 address: 192.0.2.1
***** USAGE FAULT *****
  Unaligned memory access
***** Hardware exception *****
Current thread ID = 0x80002a40
Faulting instruction address = 0x600198d6
Fatal fault in ISR! Spinning...
~~~

pfalcon commented:
~~~
600198d0 <_handle_obj_poll_events>:
        return list->head == list;
600198d0:       6803            ldr     r3, [r0, #0]
        if (!sys_dlist_is_empty(list)) {
600198d2:       4298            cmp     r0, r3
600198d4:       d007            beq.n   600198e6 <_handle_obj_poll_events+0x16>
        node->prev->next = node->next;
600198d6:       e893 0005       ldmia.w r3, {r0, r2}
~~~

I saw a crash in similar place lately, but don't remember where exactly...

pfalcon commented:
Yeah, that was logging: https://github.com/zephyrproject-rtos/zephyr/pull/12327#issuecomment-452421262

pfalcon commented:
> Yeah, that was logging: #12327 (comment)

With logging disabled, pings still don't work, the same situation as described in https://github.com/zephyrproject-rtos/zephyr/issues/11586#issuecomment-456180183

agansari committed:
mimxrt1050_evk: eth_mcux: prevent PHY NANDTree

Force Operation Mode Strap Override register to disable NANDTree. This
is due to some users reporting PHY entering NANDTree.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

MaureenHelm commented on boards/arm/mimxrt1050_evk/mimxrt1050_evk.dts:
```diff
@@ -33,7 +33,11 @@
 #elif defined(CONFIG_CODE_HYPERFLASH)
 		zephyr,flash = &hyperflash0;
 #endif
+#if defined(CONFIG_DEVICE_CONFIGURATION_DATA)
```
This logic seems backwards to me. I think the DCD should be enabled as a result of someone choosing to link data into SDRAM.

agansari replied:
@MaureenHelm Does DCD work for QSPI version of the board?

MaureenHelm replied:
It should

MaureenHelm reviewed (commented):
> NXP skd engineers can make this change, but i'm not sure it completely fixes the issue.

If we agree this change is needed, we can make the change in ext/ and feed it back to the NXP SDK team to get it into the next SDK release.

agansari commented:
@MaureenHelm this patch looks stable and there are no more coherency issues, think it's ready for merge.
Regarding fsl_enet.c barriers, they are required when used with caching memories.

agansari committed:
mimxrt1050_evk: default external flash and sdram

Moved from ITCM and DTCM to HyperFlash and SDRAM.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
mimxrt1050_evk: cache enable

Cacheing mechanisms enabled and non-cache area added.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -319,6 +327,7 @@ static void eth_mcux_phy_event(struct eth_context *context)
 			context->link_up = link_up;
 			context->phy_state = eth_mcux_phy_state_read_duplex;
 			net_eth_carrier_on(context->iface);
+			k_busy_wait(USEC_PER_MSEC);
```
Sleep will do, sorry copy-pasta from pinmux.

agansari committed:
mimxrt1050_evk: eth_mcux: driver fixup

Boot PHY initialization timeout, caching mechanism fixes and networking
buffer descriptors moved to no cache section. Enabled cache management
in networking driver and manual barriers.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari commented on drivers/ethernet/Kconfig.mcux:
```diff
@@ -7,6 +7,8 @@
 	bool "MCUX Ethernet driver"
 	depends on HAS_MCUX_ENET
+	select NOCACHE_MEMORY if HAS_MCUX_CACHE
+	select ARM_MPU if CPU_CORTEX_M7
```
@MaureenHelm done

MaureenHelm commented:
recheck
